File: AOC24/day1/main.py
from collections import Counter


def part1(s):
    
    # Setting up arrays
    ans = 0 
    left = []
    right = []

    # Convert the strings from text file to ints.
    # Placeing the left and right column numbers in the apporiate array.
    for l in s.split("\n"):
        l = [int(x) for x in l.split()]  # Convert each line into integers
        left.append(l[0])  # First number goes to `left`
        right.append(l[1])  # Second number goes to `right`
    ## Sorting array
    left.sort()
    right.sort()
    
    # Calculating the distance
    for i in range(len(left)):
        ans += abs((left[i] - right[i]))
    
    print("Answer for Part 1",ans)
    
def part2(s):

    left = []
    right = []

    # Convert the strings from text file to ints.
    # Placeing the left and right column numbers in the apporiate array.
   
    for l in s.split("\n"):
        l = [int(x) for x in l.split()]  
        left.append(l[0])  
        right.append(l[1])  

    # Create a Counter for `right` to count occurrences of each number
    right_counter = Counter(right)

    similarity_score = 0

    # Calculate the similarity score
    for l in left:
        similarity_score += l * right_counter[l]  # Multiply number by its frequency in `right`

    print("Answer for part 2:",similarity_score)




with open('/Users/ivancandelero/Desktop/AOC24/day1/input.txt') as file:
        s = file.read().strip()
part1(s)
part2(s) 

File: AOC24/day1/main.rs
use std::fs;
use std::collections::HashMap;


fn main() {
    
    // Setup ans variable. 
    let mut ans = 0;

    // Define the file path
    let file_path = "/Users/ivancandelero/Desktop/AOC24/day1/input.txt";
    let trimmed_path = file_path.trim();

    // Read file 
    let contents = fs::read_to_string(trimmed_path)
        .expect("Failed to read the file");

    // Setup arrays left and right. 
    let mut left: [i32; 1000] = [0; 1000];
    let mut right: [i32; 1000] = [0; 1000];

    // Read the lines and fill in arrays
    for (i, line) in contents.lines().enumerate() {
        if i >= 1000 {
            // Prevent overflow if the file has more than 1000 lines
            break;
        }

        // Split the line into two numbers
        let numbers: Vec<&str> = line.split_whitespace().collect();

        // Filling up arrays with the values by the column.         
        if numbers.len() == 2 {
            // Parse the numbers and assign to the arrays
            left[i] = numbers[0].parse::<i32>().expect("Invalid number in left column");
            right[i] = numbers[1].parse::<i32>().expect("Invalid number in right column");
        }
    }

    
    // Sorting Arrays 
    left.sort();
    right.sort();

    // Calculating Part 1 answer
    for i in 0..left.len() {      
        // Finding the distance of the numbers. 
        ans += (left[i] - right[i]).abs();
    }

   
    // Use a HashMap to count occurrences of each number in the right list
    let mut right_counts = HashMap::new();
    for &num in &right {
        *right_counts.entry(num).or_insert(0) += 1;
    }

    // Calculate the similarity score for Part 2
    let mut similarity_score = 0;
    for &num in &left {
        if let Some(&count) = right_counts.get(&num) {
            similarity_score += num * count;
        }
    }

    // Answers 
    println!("Answer in Part 1: {}", ans);
    println!("Answer in Part 2 {}", similarity_score);

}
